lib/logstash/outputs/s3.rb

```diff
@@ -401,26 +401,38 @@ def configure_upload_workers
       Stud::Task.new do
         LogStash::Util::set_thread_name("<S3 upload worker #{worker_id}")
 
-        while true do
+        continue = true
+        while continue do
           @logger.debug("S3: upload worker is waiting for a new file to upload.", :worker_id => worker_id)
 
-          upload_worker
+          continue = upload_worker
         end
       end
     end
   end
 
   private
   def upload_worker
-    file = @upload_queue.deq
+    file = nil
+    begin
+      file = @upload_queue.deq
 
-    case file
-      when LogStash::ShutdownEvent
+      if file.is_a? LogStash::ShutdownEvent
```

Comment:
Oops, I missed that in the first review, I think we should compare with the value of `LogStash::SHUTDOWN` instead of the class type.